File: wagerproof-ios-native/Wagerproof/Features/GameWidgets/PickSignalPills.swift
import SwiftUI
import WagerproofDesign

/// The "Pick Signals" row that sits under a market card's recommendation.
///
/// One pill per signal, tappable to open the signal's backtest sheet. Shared by
/// the CFB and NFL game sheets — both sports read the same `signal_defs` shape,
/// so the pill takes plain values rather than a per-sport model.
///
/// ## Why one header instead of two groups
/// The old layout printed "Supports this pick" and "Contradicts this pick" as
/// separate headed groups. That buried the only thing a reader has to do here —
/// tap one — under a taxonomy, and it read as two unrelated lists. Now there is
/// a single labelled affordance, and stance is carried by the pill itself:
/// amber triangle = argues against the pick, tier-coloured circle = backs it.
/// The legend under the header states that in words so the colour isn't load
/// bearing on its own.

/// One signal as it appears on a card.
struct PickSignalPillModel: Identifiable, Equatable {
    let id: String
    let title: String
    /// Argues AGAINST the card's pick. Drives the amber tint and the triangle.
    let contradicts: Bool
    /// Conviction-tier colour, used when the signal supports the pick.
    let tint: Color

    static func == (a: PickSignalPillModel, b: PickSignalPillModel) -> Bool {
        a.id == b.id && a.title == b.title && a.contradicts == b.contradicts
    }
}

/// Wrapping pill layout — `HStack` truncates rather than wrapping, and a
/// `LazyVGrid` forces a fixed column width that leaves long signal names
/// scaled down next to short ones.
struct SignalFlowLayout: Layout {
    var spacing: CGFloat = 7

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var x: CGFloat = 0
        var rowHeight: CGFloat = 0
        var total: CGFloat = 0

        for sub in subviews {
            let size = sub.sizeThatFits(.unspecified)
            if x > 0 && x + size.width > maxWidth {
                total += rowHeight + spacing
                x = 0
                rowHeight = 0
            }
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
        total += rowHeight
        return CGSize(width: maxWidth.isFinite ? maxWidth : x, height: total)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX
        var y = bounds.minY
        var rowHeight: CGFloat = 0

        for sub in subviews {
            let size = sub.sizeThatFits(.unspecified)
            if x > bounds.minX && x + size.width > bounds.maxX {
                x = bounds.minX
                y += rowHeight + spacing
                rowHeight = 0
            }
            sub.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
    }
}

/// A single Liquid Glass signal pill.
///
/// `.interactive()` glass gives the system's own touch refraction on iOS 26; the
/// scale + shadow on press is what carries that feedback on the fallback path,
/// so the pill still feels live on iOS < 26.
struct PickSignalPill: View {
    let model: PickSignalPillModel
    let action: () -> Void

    private var tint: Color { model.contradicts ? Color.appAccentAmber : model.tint }
    private var icon: String {
        model.contradicts ? "exclamationmark.triangle.fill" : "exclamationmark.circle.fill"
    }

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                Image(systemName: icon)
                    .font(.system(size: 11, weight: .black))
                    .foregroundStyle(tint)
                Text(model.title)
                    .font(.system(size: 12, weight: .heavy))
                    .foregroundStyle(Color.appTextPrimary)
                    .lineLimit(1)
            }
            .padding(.horizontal, 13)
            .padding(.vertical, 9)
            // Colour rides the ICON, not the fill. A 0.28 tint plus a hard
            // border rendered as a solid mustard button — the glass was doing
            // nothing you could see. Supporting pills take plain glass so the
            // amber ones are the only coloured thing in the row.
            .modifier(SignalPillSurface(tint: model.contradicts ? tint : nil))
            .contentShape(Capsule())
        }
        .buttonStyle(SignalPillButtonStyle(tint: tint))
        .accessibilityLabel(
            "\(model.title). \(model.contradicts ? "Contradicts this pick" : "Supports this pick"). Tap for backtest."
        )
    }
}

/// The pill's glass surface. `tint` is nil for supporting signals, which take
/// plain glass — only the contradicting ones carry amber, so the warning reads
/// at a glance instead of competing with a row of coloured chips.
///
/// A `ViewModifier` rather than an inline `@ViewBuilder` branch because
/// `liquidGlassBackground` is generic over the shape and the two arities return
/// different opaque types, which won't unify inside one `if`.
private struct SignalPillSurface: ViewModifier {
    let tint: Color?

    func body(content: Content) -> some View {
        if let tint {
            content
                .liquidGlassBackground(in: Capsule(), tint: tint.opacity(0.16), interactive: true)
                .overlay(Capsule().strokeBorder(tint.opacity(0.45), lineWidth: 1))
        } else {
            content
                .liquidGlassBackground(in: Capsule(), interactive: true)
                .overlay(Capsule().strokeBorder(Color.appBorder.opacity(0.35), lineWidth: 0.8))
        }
    }
}

/// Press feedback: a spring squeeze plus a tint glow, so the pill answers the
/// finger on every OS version.
private struct SignalPillButtonStyle: ButtonStyle {
    let tint: Color

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.94 : 1)
            .shadow(
                color: tint.opacity(configuration.isPressed ? 0.45 : 0),
                radius: configuration.isPressed ? 9 : 0
            )
            .animation(.spring(response: 0.26, dampingFraction: 0.62), value: configuration.isPressed)
    }
}

/// Header + wrapped pills. Renders nothing when there are no signals.
struct PickSignalsSection: View {
    let signals: [PickSignalPillModel]
    let onSelect: (PickSignalPillModel) -> Void

    private var hasContradicting: Bool { signals.contains(where: \.contradicts) }

    var body: some View {
        if !signals.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                HStack(spacing: 6) {
                    Text("PICK SIGNALS")
                        .font(.system(size: 10, weight: .black))
                        .tracking(0.6)
                        .foregroundStyle(Color.appTextSecondary)
                    Text("(tap for details)")
                        .font(.system(size: 9, weight: .heavy))
                        .foregroundStyle(Color.appTextMuted)
                }

                // Stance is otherwise colour-only, which fails for anyone who
                // can't separate amber from the tier tints.
                if hasContradicting {
                    HStack(spacing: 5) {
                        Image(systemName: "exclamationmark.triangle.fill")
                            .font(.system(size: 8, weight: .black))
                            .foregroundStyle(Color.appAccentAmber)
                        Text("Amber signals argue against this pick")
                            .font(.system(size: 9, weight: .bold))
                            .foregroundStyle(Color.appTextMuted)
                    }
                }

                LiquidGlassMergeContainer(spacing: 12) {
                    SignalFlowLayout(spacing: 7) {
                        ForEach(signals) { signal in
                            PickSignalPill(model: signal) { onSelect(signal) }
                        }
                    }
                }
            }
        }
    }
}

File: wagerproof-android-native/core/design/src/main/kotlin/com/wagerproof/core/design/components/PickSignalPills.kt
package com.wagerproof.core.design.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import com.wagerproof.core.design.icons.AppIcon
import com.wagerproof.core.design.tokens.AppColors

/**
 * The "Pick Signals" row under a market card's recommendation.
 *
 * Port of `Wagerproof/Features/GameWidgets/PickSignalPills.swift` and
 * `src/features/games/detail/signals/PickSignalsSection.tsx`.
 *
 * ## Why one header instead of two groups
 * The old layout printed "Supports this pick" and "Contradicts this pick" as
 * separate headed groups, which buried the only thing a reader has to do here —
 * open one — under a taxonomy. There is now a single labelled affordance, with
 * stance carried by the pill: amber triangle argues against the pick, neutral
 * circle backs it. The legend says so in words so colour isn't load bearing.
 */
data class PickSignalPillModel(
    val id: String,
    val title: String,
    /** Argues AGAINST the card's pick — drives the amber tint and the triangle. */
    val contradicts: Boolean,
)

@Composable
fun PickSignalsSection(
    signals: List<PickSignalPillModel>,
    onSelect: (PickSignalPillModel) -> Unit,
    modifier: Modifier = Modifier,
    /** Game-level lists aren't attached to a pick, so they name themselves. */
    title: String = "PICK SIGNALS",
) {
    if (signals.isEmpty()) return

    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.Bottom,
        ) {
            Text(
                title,
                color = AppColors.appTextSecondary,
                fontSize = 10.sp,
                fontWeight = FontWeight.Black,
            )
            Text(
                "(tap for details)",
                color = AppColors.appTextMuted,
                fontSize = 9.sp,
                fontWeight = FontWeight.Black,
            )
        }

        // Stance is otherwise colour-only, which fails for anyone who can't
        // separate amber from the neutral pills.
        if (signals.any { it.contradicts }) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(5.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                AppIcon.fromSystemName("exclamationmark.triangle.fill")?.let {
                    Icon(
                        it.imageVector,
                        null,
                        tint = AppColors.appAccentAmber,
                        modifier = Modifier.size(9.dp),
                    )
                }
                Text(
                    "Amber signals argue against this pick",
                    color = AppColors.appTextMuted,
                    fontSize = 9.sp,
                    fontWeight = FontWeight.Bold,
                )
            }
        }

        FlowRow(
            horizontalArrangement = Arrangement.spacedBy(7.dp),
            verticalArrangement = Arrangement.spacedBy(7.dp),
        ) {
            signals.forEach { PickSignalPill(it) { onSelect(it) } }
        }
    }
}

@Composable
private fun PickSignalPill(model: PickSignalPillModel, onTap: () -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    // Compose has no `.interactive()` glass, so the squeeze is what makes the
    // pill answer the finger.
    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.94f else 1f,
        animationSpec = spring(dampingRatio = 0.62f, stiffness = 900f),
        label = "signalPillPress",
    )

    val tint = if (model.contradicts) AppColors.appAccentAmber else null
    val icon = if (model.contradicts) "exclamationmark.triangle.fill" else "exclamationmark.circle.fill"

    Row(
        Modifier
            .scale(scale)
            // Colour rides the ICON, not the fill: supporting pills take plain
            // glass so the amber ones are the only coloured thing in the row.
            .liquidGlassCapsule(tint = tint)
            .then(
                if (tint != null) {
                    Modifier.border(1.dp, tint.copy(alpha = 0.45f), CircleShape)
                } else {
                    Modifier.border(0.8.dp, AppColors.appBorder.copy(alpha = 0.35f), CircleShape)
                }
            )
            .clickable(interactionSource = interaction, indication = null, onClick = onTap)
            // Stance has to be spoken, not just tinted — matches the iOS
            // accessibility label and the web aria-label.
            .semantics {
                contentDescription = "${model.title}. " +
                    (if (model.contradicts) "Contradicts this pick" else "Supports this pick") +
                    ". Tap for backtest."
            }
            .padding(horizontal = 13.dp, vertical = 9.dp),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        AppIcon.fromSystemName(icon)?.let {
            Icon(
                it.imageVector,
                null,
                tint = tint ?: AppColors.appTextSecondary,
                modifier = Modifier.size(11.dp),
            )
        }
        Text(
            model.title,
            color = AppColors.appTextPrimary,
            fontSize = 12.sp,
            fontWeight = FontWeight.Black,
            maxLines = 1,
        )
    }
}
